### PR TITLE
Remove hacks from e2e tests

### PR DESCRIPTION
**Pull request by pradeepitm12**
- fixes command to return the correct exit code
- removes hacks from e2e testes
- improves existing e2e tests
- more test coverage

Signed-off-by: Pradeep Kumar <pradkuma@redhat.com>

# Changes

# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [ ] Includes [tests](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if functionality changed/added)
- [ ] Includes [docs](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if user facing)
- [ ] Commit messages follow [commit message best practices](https://github.com/tektoncd/community/blob/master/standards.md#commit-messages)

_See [the contribution guide](https://github.com/tektoncd/cli/blob/master/CONTRIBUTING.md)
for more details._

# Release Notes

```
release-note
```

### CODE REVIEW

File: pkg/cmd/pipeline/describe.go [hunk 1]
```diff
@@ -63,7 +63,8 @@ func printPipelineDescription(out io.Writer, p cli.Params, pname string) error {
 
 	pipeline, err := cs.Tekton.TektonV1alpha1().Pipelines(p.Namespace()).Get(pname, metav1.GetOptions{})
 	if err != nil {
-		return err
+		fmt.Fprintf(out, emptyMsg)
```
Comment (sthaha) [inline pkg/cmd/pipeline/describe.go]: I am tempted to use `stderr` to write this message so that `out` is empty

Comment (sthaha) [reply pkg/cmd/pipeline/describe.go]: @chmouel @vdemeester wdyt?

File: pkg/cmd/pipeline/describe.go [hunk 2]
```diff
@@ -63,7 +63,8 @@ func printPipelineDescription(out io.Writer, p cli.Params, pname string) error {
 
 	pipeline, err := cs.Tekton.TektonV1alpha1().Pipelines(p.Namespace()).Get(pname, metav1.GetOptions{})
 	if err != nil {
-		return err
```
Comment (sthaha) [inline pkg/cmd/pipeline/describe.go]: this must be an error because user asked to display details of a pipeline **which does not exist**

File: pkg/cmd/pipeline/list.go [hunk 3]
```diff
@@ -81,7 +81,8 @@ func printPipelineDetails(out io.Writer, p cli.Params) error {
 	}
 
 	if len(ps.Items) == 0 {
-		return fmt.Errorf(emptyMsg)
+		fmt.Fprintf(out, emptyMsg)
```
Comment (sthaha) [inline pkg/cmd/pipeline/list.go]: how about `stderr` instead of `stdout` ?

File: pkg/cmd/pipeline/describe.go (hunk 1 above)
Comment (vdemeester) [reply pkg/cmd/pipeline/describe.go]: yes make sense to me

File: pkg/cmd/pipeline/list_test.go [hunk 4]
```diff
@@ -24,12 +24,12 @@ func TestPipelinesList_empty(t *testing.T) {
...
-	_, err := test.ExecuteCommand(pipeline, "list", "-n", "foo")
-	if err == nil {
-		t.Errorf("No errors was defined")
+	output, err := test.ExecuteCommand(pipeline, "list", "-n", "foo")
+	if err != nil {
+		t.Errorf("Unexpected error: %v", err)
 	}
 
-	if d := cmp.Diff(emptyMsg, err.Error()); d != "" {
+	if d := cmp.Diff(emptyMsg, output); d != "" {
```
Comment (sthaha) [inline pkg/cmd/pipeline/list_test.go]: nit: you can use tu.AssertOutput()

File: pkg/cmd/pipelinerun/describe.go [hunk 5]
```diff
@@ -18,14 +18,15 @@ import (
 	"fmt"
 	"io"
 
+	"text/tabwriter"
```
Comment (sthaha) [inline pkg/cmd/pipelinerun/describe.go]: was this change intentional? seems unrelated

File: pkg/cmd/pipelinerun/describe.go [hunk 6]
```diff
@@ -68,12 +69,13 @@ tkn pr desc foo -n bar",
 func printPipelineRunDescription(out io.Writer, prname string, p cli.Params) error {
 	cs, err := p.Clients()
 	if err != nil {
-		return fmt.Errorf("Failed to create tekton client\n")
+		return fmt.Errorf("failed to create tekton client")
 	}
 
 	pr, err := cs.Tekton.TektonV1alpha1().PipelineRuns(p.Namespace()).Get(prname, metav1.GetOptions{})
 	if err != nil {
```
Comment (sthaha) [inline pkg/cmd/pipelinerun/describe.go]: `describe` must fail with non-zero exit code if the pipelinerun does not exist

File: pkg/cmd/pipelinerun/list.go [hunk 7]
```diff
@@ -119,7 +119,7 @@ func list(p cli.Params, pipeline string) (*v1alpha1.PipelineRunList, error) {
 
 func printFormatted(out io.Writer, prs *v1alpha1.PipelineRunList, c clockwork.Clock) error {
 	if len(prs.Items) == 0 {
-		fmt.Fprintln(out, msgNoPRsFound)
+		fmt.Fprintf(out, emptyMsg)
```
Comment (sthaha) [inline pkg/cmd/pipelinerun/list.go]: `stderr` instead of `out`?

File: pkg/cmd/pipelinerun/log_reader.go [hunk 8]
```diff
@@ -50,7 +51,8 @@ func (lr *LogReader) Read() (<-chan Log, <-chan error, error) {
 	tkn := lr.Clients.Tekton
 	pr, err := tkn.TektonV1alpha1().PipelineRuns(lr.Ns).Get(lr.Run, metav1.GetOptions{})
 	if err != nil {
-		return nil, nil, fmt.Errorf("%s : %s", msgPRNotFoundErr, err)
+		fmt.Fprintf(lr.Stream.Out, emptyMsg)
```
Comment (sthaha) [inline pkg/cmd/pipelinerun/log_reader.go]: logs must fail if the `pipelinerun` isn't found.

File: pkg/cmd/task/list.go [hunk 9]
```diff
@@ -73,12 +73,13 @@ func printTaskDetails(out io.Writer, p cli.Params) error {
 
 	tasks, err := listTaskDetails(cs.Tekton, p.Namespace())
 	if err != nil {
-		fmt.Fprintf(os.Stderr, "Failed to list tasks from %s namespace\n", p.Namespace())
-		return err
+		fmt.Fprintf(out, emptyMsg)
```
Comment (sthaha) [inline pkg/cmd/task/list.go]: same; must fail if the object does not exist

File: pkg/cmd/taskrun/log_reader.go [hunk 10]
```diff
@@ -58,7 +60,8 @@ func (lr *LogReader) Read() (<-chan Log, <-chan error, error) {
 	tkn := lr.Clients.Tekton
 	tr, err := tkn.TektonV1alpha1().TaskRuns(lr.Ns).Get(lr.Run, metav1.GetOptions{})
 	if err != nil {
-		return nil, nil, fmt.Errorf("%s : %s", msgTRNotFoundErr, err)
+		fmt.Fprintf(lr.Stream.Out, msgTRNotFoundErr)
```
Comment (sthaha) [inline pkg/cmd/taskrun/log_reader.go]: same; must fail

File: test/e2e-tests.sh [hunk 11]
```diff
@@ -92,16 +91,15 @@ tkn() {
...
+run_test  "list pipelines" tkn pipeline list
+run_test  "describe pipeline" tkn pipeline describe foo
+run_test  "list pipelinerun" tkn pipelinerun list
+run_test  "describe pipelinerun" tkn pipelinerun describe foo
+run_test  "show logs" tkn pipelinerun logs foo
+run_test  "list task" tkn task list
+run_test  "show logs" tkn taskrun logs foo
+run_test  "list taskrun" tkn taskrun list
 
-# 0| dt"| ut"d$ |0|irun_test| p|li lxxt|d$|0t"| lldwdw
```
Comment (sthaha) [inline test/e2e-tests.sh]: 🤗

File: pkg/cmd/pipelinerun/describe.go (hunk 5 above)
Comment (pradeepitm12) [reply pkg/cmd/pipelinerun/describe.go]: My vs code put it at the top, will fix it.

File: pkg/cmd/task/list.go (hunk 9 above)
Comment (pradeepitm12) [reply pkg/cmd/task/list.go]: **tkn task list** does not take a particular name, so we could only print the message like other list commands.

File: test/e2e-tests.sh [hunk 12]
```diff
@@ -62,6 +62,22 @@ run_test() {
...
 
+# runs the test which should fail
+# to run a fail test
+#   run_fail_test "describe pipeline" ./tkn pipeline describe foo -n default
+run_fail_test() {
+  local desc="$1"; shift
+  local failed=0
+  echo "Running fail $@"
+  $@
+  ecode=$?
```
Comment (sthaha) [inline test/e2e-tests.sh]: could use the same pattern as below 

```
 local failed=0
  $@ || failed=1
  (( failed )) ||fail_test
```

File: test/e2e-tests.sh [hunk 13]
```diff
@@ -79,29 +94,22 @@ ci_run && {
...
-run_test  "describe pipeline" SKIP tkn pipeline describe foo
-run_test  "list pipelinerun" SKIP tkn pipelinerun list
-run_test  "describe pipelinerun" SKIP tkn pipelinerun describe foo
-run_test  "show logs" SKIP tkn pipelinerun logs foo
-run_test  "list task" SKIP tkn task list
-run_test  "show logs" SKIP tkn taskrun logs foo
-run_test  "list taskrun" SKIP tkn taskrun list
+# command before creation of resources
+run_test  "list pipelines" tkn pipeline list
+run_fail_test  "describe pipeline" tkn pipeline describe foo
```
Comment (sthaha) [inline test/e2e-tests.sh]: describe pipeline that does not exist

File: pkg/cmd/pipelinerun/describe.go (hunk 5 above)
Comment (chmouel) [reply pkg/cmd/pipelinerun/describe.go]: it should respect  whatever we have in `go fmt` and `goimports` isnt it ?

File: pkg/cmd/pipelinerun/list.go [hunk 14]
```diff
@@ -60,7 +59,7 @@ tkn pr list -n foo",
 			prs, err := list(p, pipeline)
 			if err != nil {
 				fmt.Fprintf(os.Stderr, "Failed to list pipelineruns from %s namespace \n", p.Namespace())
-				return err
```
Comment (hrishin) [inline pkg/cmd/pipelinerun/list.go]: It should return `error`?

File: pkg/cmd/taskrun/list.go [hunk 15]
```diff
@@ -60,7 +59,7 @@ tkn pr list -n foo \n",
 			trs, err := list(p, task)
 			if err != nil {
 				fmt.Fprintf(os.Stderr, "Failed to list taskruns from %s namespace \n", p.Namespace())
-				return err
+				return nil
```
Comment (hrishin) [inline pkg/cmd/taskrun/list.go]: This should also return an error?

File: pkg/cmd/taskrun/log_reader.go [hunk 16]
```diff
@@ -16,6 +16,7 @@ package taskrun
 
 import (
 	"fmt"
+
```
Comment (hrishin) [inline pkg/cmd/taskrun/log_reader.go]: :+1:

File: test/e2e-tests.sh [hunk 17]
```diff
@@ -79,29 +90,22 @@ ci_run && {
...
-run_test  "list pipelinerun" SKIP tkn pipelinerun list
-run_test  "describe pipelinerun" SKIP tkn pipelinerun describe foo
-run_test  "show logs" SKIP tkn pipelinerun logs foo
-run_test  "list task" SKIP tkn task list
-run_test  "show logs" SKIP tkn taskrun logs foo
-run_test  "list taskrun" SKIP tkn taskrun list
+# command before creation of resources
+run_test  "list pipelines" tkn pipeline list
+run_fail_test  "describe pipeline that does not exist" tkn pipeline describe foo
+run_test  "list pipelinerun" tkn pipelinerun list
```
Comment (hrishin) [inline test/e2e-tests.sh]: also need 
`run_fail_test "list pipelinerun for pipeline foo that doesn't not exist" tkn pipelinerun list foo`?

Comment (hrishin) [reply test/e2e-tests.sh]: same behavior for `taskrun list` is applicable

File: pkg/cmd/pipelinerun/list.go (hunk 14 above)
Comment (pradeepitm12) [reply pkg/cmd/pipelinerun/list.go]: If the resource is not found then we get the below error
Error: the server could not find the requested resource (get pipelineruns.tekton.dev)

File: pkg/cmd/pipelinerun/list.go [hunk 18]
```diff
@@ -90,6 +92,9 @@ tkn pr list -n foo",
 func list(p cli.Params, pipeline string) (*v1alpha1.PipelineRunList, error) {
 	cs, err := p.Clients()
 	if err != nil {
+		fmt.Println("Test Logs ******* --- > error while p.Clients()")
```
Comment (sthaha) [inline pkg/cmd/pipelinerun/list.go]: debug message?

File: pkg/cmd/pipelinerun/list.go [hunk 19]
```diff
@@ -103,6 +108,9 @@ func list(p cli.Params, pipeline string) (*v1alpha1.PipelineRunList, error) {
 	prc := cs.Tekton.TektonV1alpha1().PipelineRuns(p.Namespace())
 	prs, err := prc.List(options)
 	if err != nil {
+		fmt.Println("Test Logs ******* --- > error while prc.List(options)")
```
Comment (sthaha) [inline pkg/cmd/pipelinerun/list.go]: debug message?

File: pkg/cmd/task/list.go [hunk 20]
```diff
@@ -73,23 +76,23 @@ func printTaskDetails(out io.Writer, p cli.Params) error {
 
 	tasks, err := listTaskDetails(cs.Tekton, p.Namespace())
 	if err != nil {
-		fmt.Fprintf(os.Stderr, "Failed to list tasks from %s namespace\n", p.Namespace())
-		return err
+		fmt.Fprintln(s.Err, emptyMsg)
```
Comment (sthaha) [inline pkg/cmd/task/list.go]: why are we ignoring this err?

File: pkg/cmd/task/list_test.go [hunk 21]
```diff
@@ -22,24 +22,23 @@ import (
...
+func TestTaskList_Empty(t *testing.T) {
 	cs, _ := pipelinetest.SeedTestData(pipelinetest.Data{})
 	p := &test.Params{Tekton: cs.Pipeline}
 
 	task := Command(p)
-	_, err := test.ExecuteCommand(task, "list", "-n", "foo")
-	if err == nil {
+	output, err := test.ExecuteCommand(task, "list", "-n", "foo")
+	if err != nil {
 		t.Errorf("No errors was defined")
```
Comment (sthaha) [inline pkg/cmd/task/list_test.go]: no error was defined?

File: pkg/cmd/pipelinerun/list.go (hunk 18 above)
Comment (pradeepitm12) [reply pkg/cmd/pipelinerun/list.go]: Removed.

File: pkg/cmd/pipelinerun/list.go (hunk 19 above)
Comment (pradeepitm12) [reply pkg/cmd/pipelinerun/list.go]: Removed.

File: pkg/cmd/task/list_test.go (hunk 21 above)
Comment (pradeepitm12) [reply pkg/cmd/task/list_test.go]: Corrected.

File: test/e2e-tests.sh (hunk 17 above)
Comment (pradeepitm12) [reply test/e2e-tests.sh]: Added.

File: pkg/cmd/task/list.go (hunk 20 above)
Comment (pradeepitm12) [reply pkg/cmd/task/list.go]: addressed

File: pkg/cmd/pipelinerun/list.go [hunk 22]
```diff
@@ -32,7 +31,7 @@ import (
 )
 
 const (
-	msgNoPRsFound = "No pipelineruns found."
+	emptyMsg = "No pipelineruns found"
```
[the remark this answers is not in the document]
Comment (sthaha) [reply pkg/cmd/pipelinerun/list.go]: @hrishin I am not sure if I agree with the `nit`. Where/Who has defined this convention?

File: pkg/cmd/pipeline/list.go [hunk 23]
```diff
@@ -76,15 +79,16 @@ func printPipelineDetails(out io.Writer, p cli.Params) error {
 	ps, prs, err := listPipelineDetails(cs.Tekton, p.Namespace())
 
 	if err != nil {
-		fmt.Fprintf(os.Stderr, "Failed to list pipelines from %s namespace\n", p.Namespace())
+		fmt.Fprintf(s.Err, "Failed to list pipelines from %s namespace\n", p.Namespace())
```
[the remark this answers is not in the document]
Comment (sthaha) [reply pkg/cmd/pipeline/list.go]: why do we insist on making this `lowercase`? It does not make sense to me

File: pkg/cmd/pipelinerun/list.go (hunk 22 above)
[the remark this answers is not in the document]
Comment (hrishin) [reply pkg/cmd/pipelinerun/list.go]: @sthaha the how is this is different https://github.com/tektoncd/cli/pull/127/files#diff-6483c1d87e911113778e9b15ef9e560bR71? :thinking:

[the remark this answers is not in the document]
Comment (hrishin) [reply pkg/cmd/pipelinerun/list.go]: okay got it :+1: , my bad then :sweat_smile:
